### PR TITLE
Async Responses

### PR DESCRIPTION
**Pull request by DingoEatingFuzz**
PR for issue https://github.com/trek/pretender/issues/41

- Makes responses asynchronous by default
- Responses resolve on the next run of the event loop by default
- Responses can be made synchronous by passing `false` for the async argument when specifying a route
```js
pretender.get('/sync/thing', handlerFn, false)
```
- Responses can be set to resolve after a specified time by passing in a number for the async argument
```js
pretender.get('/timed/thing', handlerFn, 300) // in ms just like setTimeout
```
- Responses can be set to never automatically resolve by passing in `true` for the async argument
```js
pretender.get('/manual/thing', handlerFn, true);
// ...
pretender.resolve(request);
```
- Responses can be set to conditionally resolve by passing in a function for the async argument
```js
pretender.get('/maybe/async/thing', handlerFn, function() { return isAsync ? 0 : false; })
```
- A helper by the name of `pretender.requiresManualResolution` exists to check if a request will, well, require manual resolution.

### CODE REVIEW

File: pretender.js [hunk 1]
```diff
@@ -178,24 +179,66 @@ Pretender.prototype = {
...
+      });
+
+      if (strategy !== true) {
+        setTimeout(function() {
+          pretender.resolve(request);
+        }, typeof strategy === 'number' ? strategy : 0);
+      }
+    }
+  },
+  resolve: function resolve(request) {
```
Comment (trek) [inline pretender.js]: What's `fn`?

Comment (DingoEatingFuzz) [inline pretender.js]: `fn` is an ambiguous variable name :stuck_out_tongue: 

It's the callback that handles the response. I will rename it to something better.

**Comment by DingoEatingFuzz** [comment]
@trek, this is ready for review again (not sure if new commits emit notifications).

File: pretender.js [hunk 2]
```diff
@@ -21,6 +21,7 @@ function Pretender(maps){
   this.handledRequests = [];
   this.passthroughRequests = [];
   this.unhandledRequests = [];
+  this.requestRefs = [];
```
Comment (trek) [inline pretender.js]: This existing code doesn't tend to use abbreviations. This ref says "Yellow Card"

File: pretender.js [hunk 3]
```diff
@@ -178,24 +180,65 @@ Pretender.prototype = {
...
+            pretender = this;
 
-        this.handledRequest(verb, path, request);
+        this.handleResponse(request, async, function() {
+          request.respond(status, headers, body);
+          pretender.handledRequest(verb, path, request);
+        });
       } catch (error) {
         this.erroredRequest(verb, path, request, error);
+        pretender.resolve(request);
```
Comment (trek) [inline pretender.js]: I'd prefer to keep this as `this`.

**Comment by trek** [comment]
Looks good other than a few style choices. I think once those are fixed, squash into a single feature commit for merge.

File: pretender.js (hunk 2 above)
Comment (DingoEatingFuzz) [inline pretender.js]: Change it to `requestReferees`, got it :trollface:

**Comment by DingoEatingFuzz** [comment]
All squashed up.

**Comment by trek** [comment]
Can you update the README? The docs are probably all wrong now.

**Comment by DingoEatingFuzz** [comment]
Yep, working on those changes now.

**Pull request merged**